### PR TITLE
[Merged by Bors] - feat(SetTheory/Ordinal/Nimber): nim addition of natural numbers is XOR

### PR DESCRIPTION
**Pull request by vihdzp**
We also introduce the notation `∗o` for `Ordinal.toNimber o`. Note that for `n : ℕ`, `∗n` is **not** the same as `NatCast`, as e.g. `∗2 ≠ ↑2 = 0`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/SetTheory/Ordinal/Nimber.lean [hunk 1]
```diff
@@ -357,4 +361,32 @@ theorem add_trichotomy {a b c : Nimber} (h : a + b + c ≠ 0) :
   · rw [← hx'] at hx
     exact Or.inr <| Or.inr hx
 
+theorem add_nat (a b : ℕ) : ∗a + ∗b = ∗(a ^^^ b) := by
```
Comment (Command-Master) [inline Mathlib/SetTheory/Ordinal/Nimber.lean:364]: Is it possible to deduce this from `grundyValue_nim_add_nim`, or maybe the other way around?

Comment (vihdzp) [reply Mathlib/SetTheory/Ordinal/Nimber.lean:364]: We should be doing this the other way around. I plan to redefine `grundyValue` as a nimber in the near future, and to prove the more general result `grundyValue (nim a + nim b) = ∗a + ∗b`.

**Comment by b-mehta** [comment]
> Note that for n : ℕ, ∗n is not the same as NatCast, as e.g. ∗2 ≠ ↑2 = 0.

This should definitely be recorded in a docstring in this file - maybe with a comment that the axioms for `AddMonoidWithOne` are not satisfied with `natCast n := ∗n`.

Is the prefix-asterisk notation standard for this embedding?

File: Mathlib/SetTheory/Ordinal/Nimber.lean [hunk 2]
```diff
@@ -29,6 +30,10 @@ To reduce API duplication, we opt not to implement operations on `Nimber` on `Or
 isomorphisms `Ordinal.toNimber` and `Nimber.toOrdinal` allow us to cast between them whenever
 needed.
 
+We also define notation `∗o` for `Ordinal.toNimber o` in the `Nimber` namespace. Note that for
```
Comment (b-mehta) [inline Mathlib/SetTheory/Ordinal/Nimber.lean:33]:
```suggestion
## Notations

We also define notation `∗o` for `Ordinal.toNimber o` in the `Nimber` namespace. Note that for
```
and the notation section should go above the implementation notes section (https://leanprover-community.github.io/contribute/doc.html).

File: Mathlib/SetTheory/Ordinal/Nimber.lean [hunk 3]
```diff
@@ -149,42 +157,40 @@ end Nimber
...
-variable {a b c : Ordinal.{u}}
+namespace Ordinal
 
 @[simp]
 theorem toNimber_symm_eq : toNimber.symm = Nimber.toOrdinal :=
   rfl
 
 @[simp]
-theorem toNimber_toOrdinal (a : Ordinal) :  Nimber.toOrdinal (toNimber a) = a :=
+theorem toNimber_toOrdinal (a : Ordinal) :  Nimber.toOrdinal (∗a) = a :=
```
Comment (b-mehta) [inline Mathlib/SetTheory/Ordinal/Nimber.lean:169]:
```suggestion
theorem toNimber_toOrdinal (a : Ordinal) : Nimber.toOrdinal (∗a) = a :=
```

**Comment by b-mehta** [comment]
> The notation is standard, it's introduced in page 121 of Conway's "On Numbers And Games".

Alright, I'd like if this were recorded in the file too. A reference to this book from mathlib would surely be sensible regardless!

**Comment by vihdzp** [comment]
The reference already exists! I've added it to the docstring.

File: Mathlib/SetTheory/Ordinal/Nimber.lean (hunk 1 above)
Comment (b-mehta) [reply Mathlib/SetTheory/Ordinal/Nimber.lean:370]: I agree this is a good way round - do you want to add a TODO in this file saying to define grundyValue as a nimber?

File: Mathlib/SetTheory/Ordinal/Nimber.lean [hunk 4]
```diff
@@ -356,4 +366,33 @@ theorem add_trichotomy {a b c : Nimber} (h : a + b + c ≠ 0) :
...
     exact Or.inr <| Or.inr hx
 
+/-- Nimber addition of naturals corresponds to the bitwise XOR operation. -/
+theorem add_nat (a b : ℕ) : ∗a + ∗b = ∗(a ^^^ b) := by
+  apply le_antisymm
+  · apply add_le_of_forall_ne
+    all_goals
+      refine Nimber.rec (fun c hc ↦ ?_)
+      rw [OrderIso.lt_iff_lt] at hc
+      obtain ⟨c, rfl⟩ := Ordinal.lt_omega0.1 (hc.trans (Ordinal.nat_lt_omega0 _))
```
Comment (b-mehta) [inline Mathlib/SetTheory/Ordinal/Nimber.lean:376]: You use this pattern twice - maybe it should be a lemma? If an ordinal is <= a natural, then it's equal to a natural.

File: Mathlib/SetTheory/Ordinal/Nimber.lean [hunk 5]
```diff
@@ -356,4 +366,33 @@ theorem add_trichotomy {a b c : Nimber} (h : a + b + c ≠ 0) :
...
+      replace hc := Nat.cast_lt.1 hc
+      rw [add_nat, ne_eq, EmbeddingLike.apply_eq_iff_eq, Nat.cast_inj]
+      have := hc.ne
+    · rwa [Nat.xor_left_inj]
+    · rwa [Nat.xor_right_inj]
```
Comment (b-mehta) [inline Mathlib/SetTheory/Ordinal/Nimber.lean:381]:
```suggestion
      rw [add_nat]
      simpa using hc.ne
```
a golf, if you like.

File: Mathlib/SetTheory/Ordinal/Nimber.lean [hunk 6]
```diff
@@ -356,4 +366,33 @@ theorem add_trichotomy {a b c : Nimber} (h : a + b + c ≠ 0) :
...
+    rw [hc', Nat.cast_lt] at hc
+    obtain h | h := Nat.lt_xor_cases hc
+    · have := add_nat (c ^^^ b) b
+      rw [Nat.xor_cancel_right, ← hc', toOrdinal_toNimber, add_left_inj,
+        EquivLike.apply_eq_iff_eq, Nat.cast_inj] at this
+      exact h.ne this
+    · have := add_nat a (c ^^^ a)
+      rw [Nat.xor_comm, Nat.xor_cancel_left, ← hc', toOrdinal_toNimber, add_right_inj,
+        EquivLike.apply_eq_iff_eq, Nat.cast_inj, Nat.xor_comm] at this
+      exact h.ne this
```
Comment (b-mehta) [inline Mathlib/SetTheory/Ordinal/Nimber.lean:395]:
```suggestion
    · apply h.ne
      simpa [Nat.xor_comm, Nat.xor_cancel_left, ← hc'] using add_nat (c ^^^ b) b
    · apply h.ne
      simpa [Nat.xor_comm, Nat.xor_cancel_left, ← hc'] using add_nat a (c ^^^ a)
```
a golf, if you like

File: Mathlib/SetTheory/Ordinal/Nimber.lean [hunk 7]
```diff
@@ -356,4 +366,33 @@ theorem add_trichotomy {a b c : Nimber} (h : a + b + c ≠ 0) :
...
+termination_by (a, b)
```
Comment (b-mehta) [inline Mathlib/SetTheory/Ordinal/Nimber.lean:396]:
```suggestion
```
Lean figures this one out!

**Review by b-mehta — APPROVED** [review]

File: Mathlib/SetTheory/Ordinal/Nimber.lean (hunk 1 above)
Comment (b-mehta) [reply Mathlib/SetTheory/Ordinal/Nimber.lean:370]: Great!

File: Mathlib/SetTheory/Ordinal/Nimber.lean (hunk 4 above)
Comment (vihdzp) [reply Mathlib/SetTheory/Ordinal/Nimber.lean]: I've added `eq_nat_of_le_nat`, it does simplify the proof a bit.

File: Mathlib/SetTheory/Ordinal/Nimber.lean (hunk 5 above)
Comment (vihdzp) [reply Mathlib/SetTheory/Ordinal/Nimber.lean:384]: really nice!

File: Mathlib/SetTheory/Ordinal/Nimber.lean [hunk 8]
```diff
@@ -356,4 +369,25 @@ theorem add_trichotomy {a b c : Nimber} (h : a + b + c ≠ 0) :
...
 
+/-- Nimber addition of naturals corresponds to the bitwise XOR operation. -/
+theorem add_nat (a b : ℕ) : ∗a + ∗b = ∗(a ^^^ b) := by
+  apply le_antisymm
+  · apply add_le_of_forall_ne
+    all_goals
+      intro c hc
+      obtain ⟨c, rfl⟩ := eq_nat_of_le_nat hc.le
+      rw [OrderIso.lt_iff_lt] at hc
+      replace hc := Nat.cast_lt.1 hc
```
Comment (b-mehta) [inline Mathlib/SetTheory/Ordinal/Nimber.lean:380]:
```suggestion
      rw [OrderIso.lt_iff_lt, Nat.cast_lt] at hc
```
seems surely better (you might need a \l on the second one, I didn't check)

**Comment by b-mehta** [comment]
bors d+

File: Mathlib/SetTheory/Ordinal/Nimber.lean (hunk 8 above)
Comment (vihdzp) [reply Mathlib/SetTheory/Ordinal/Nimber.lean:380]: For whatever reason, the termination checker fails when doing this. I'm not really sure what's going on there.

Comment (b-mehta) [reply Mathlib/SetTheory/Ordinal/Nimber.lean:380]: Ah, probably the old version is kept around secretly somewhere, and the termination checker is using that as proof of termination

**Pull request closed without merging**